A user guide for the factory pattern. (#983)
Concordium/concordium.github.io: source/mainnet/smart-contracts/guides/factory-pattern.rst

.. _factory-pattern:

=======================================
Using the factory pattern on Concordium
=======================================

.. Warning::

    This guide makes use of features that are first available in protocol version 7,
    namely getting the contract name and module reference of a smart contract instance
    from a smart contract.

>     Before the protocol upgrade to version 7, these features will not work.

Reviewer: Could we link to a page where people can check when version 7 will be expected to be live (or give a tentative date here e.g. around summer).

Reviewer 2: We don't have that yet.

This process is significantly more complex than the idealised factory pattern we started with.
In particular, the user is required to sign two separate transactions, with the second one depending
on the result of the first one. If this two-step process is presented to the end user, it is likely
to cause confusion. Additionally, two-step process means that the smart contracts have to be robust
against misuse. For instance, the factory has to deal with the possibility that ``produce`` is
called with a reference to the wrong type of product contract.

Alternatives to the factory pattern
===================================

Since there are a number of disadvantages to using the factory pattern on Concordium, it is
important to consider whether a better alternative is available. One of the following alternatives
may be suitable, depending on the application.

Construct the product directly
------------------------------

If the factory does not maintain an on-going relationship with the product,
then it is generally possible to simply initialize the product entirely in
the constructor (``init_product``), removing the separate ``initialize`` operation entirely.

> In general, if the factory contract does not need to update its state
> after initializing the product, then it serves no purpose. It would be

Reviewer: If I understand correctly, you want to state that only if the `factory` stores some data in its state in the same function call when `producing` a new `product` then it has a purpose. 

Since the factory can also update its state before `initializing` the product which is the recommended pattern `check-effect-interact`, I would rephrase this to:

```suggestion
In general, if the factory contract does not need to update its state
when initializing the product, then it serves no purpose. It would be
```

----------------------------------------------

A factory contract may serve as a registry of its products.
The factory pattern enables this, as the factory is responsible for initializing each product
instance. However, the factory pattern is not in itself necessary for a contract registry.
Instead, instances could be constructed directly (as discussed above) and subsequently registered
with a registry contract, which maintains an index of the instances registered with it.
The registration process could also provide additional initialization information for the contract
being registered (such as the index it is assigned by the registry).

Use a monolithic contract
-------------------------

Instead of having separate contract instances for the factory and each product, all of these
could be combined in a single monolithic contract. In this model, each product is assigned a
virtual address, and the contract maintains a mapping from virtual addresses to the state of
each product.

For instance, a CIS2 contract can manage multiple NFTs that each have distinct token IDs.
Rather than each NFT being its own contract instance (created by a factory), they are simply

> handled as part of the state of the overall CIS2 contract. Here, the token ID actts as a

Reviewer: ```suggestion
handled as part of the state of the overall CIS2 contract. Here, the token ID acts as a
```

virtual address.

The main disadvantage of this approach is that the isolation between the states of each product
must be enforced by the contract itself. If the state becomes corrupted (due to a bug in the
contract) then all of the encapsulated products can potentially be affected. With the factory

> pattern, the runtime system of the blockchain enforces isolation. As always, the balance of risks

Reviewer 2: This argument here about isolation seems spurious because if all contracts are initialized from the same code, then corruption will likely also lead to corruption in all products.

So this is neither here nor there.

protocol version.

In this example, the factory and product contracts are defined in the same module.
Thus, to check that the module reference of the product is correct, it is sufficient to check that
it is the same as the module reference of the factory, as follows:

.. code-block:: Rust

        // We can depend upon getting the module reference for our own contract.
        let self_module_ref = host.contract_module_reference(ctx.self_address()).unwrap();
        // Check the product module is the same as our own module.
        let product_module_ref = host
            .contract_module_reference(product_address)
            .or(Err(FactoryError::NonExistentProduct))?;
        ensure_eq!(self_module_ref, product_module_ref, FactoryError::InvalidProduct);

If the product were instead defined in a different module from the factory, then the expected
module reference for the product could be hard-coded into the factory. Alternatively, if the
module rereference for the product is to be determined later, it could be passed as a parameter
when creating the ``factory`` instance.

If the module defining the product is known to only contain one smart contract, then checking the

> module reference is sufficient for identifying the code of the product smart contract. In this case,

Reviewer: ```suggestion
module reference is sufficient for identifying the code of the product smart contract. In the example smart contract,
```

People could think that `in this case` refers to the example you describe in the previous sentence.

however, both the factory and product contracts are defined in the same module, so it is necessary
to also check the contract name. This is achieved as follows:

.. code-block:: Rust

        let product_name =
            host.contract_name(product_address).or(Err(FactoryError::NonExistentProduct))?;
        ensure_eq!(product_name, PRODUCT_INIT_NAME, FactoryError::InvalidProduct);

Now the contract is known to be an instance of ``product``, the next step is to call
``initialize``. In this example, ``initialize`` takes as a parameter that is the index assigned to
it, which will be the current value of ``next_product`` in the state.

.. code-block:: Rust

        let next_product = host.state().next_product;
        host.invoke_contract(
            &product_address,
            &next_product,
            EntrypointName::new_unchecked(PRODUCT_INITIALIZE_ENTRYPOINT),
            Amount::zero(),
        )
        .or(Err(FactoryError::InitializeFailed))?;

Here, it is assumed that ``initialize`` will fail, for instance, if it is called on a product that
has previously been initialized.
It only remains to update the factory contract's state:

.. code-block:: Rust

>         let state = host.state_mut();

Reviewer: The usual smart contract best practice pattern is to do a `check-effect-interact` pattern. This means we should do the state update in the factory before calling an external contract. We can consider to change the order in this function in the factory example.

uninitialized. The full state of the product is thus represented as follows:

.. code-block:: Rust

    pub enum ProductState {
        /// The product has not yet been initialized by the factory.
        Uninitialized,
        /// The product has been initialised by the factory.
        Initialized(Product),
    }

The ``init`` function for the product will simply create a new product in the ``Uninitialized``
state:

.. code-block:: Rust

    #[init(contract = "product")]
    pub fn init(_ctx: &InitContext, _state_builder: &mut StateBuilder) -> InitResult<ProductState> {
        Ok(ProductState::Uninitialized)
    }

The ``initialize`` endpoint
^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``initialize`` method of the product first checks that the product has not already been
initialized:

.. code-block:: Rust

        let state = host.state_mut();
        ensure_eq!(*state, ProductState::Uninitialized, ProductError::AlreadyInitialized);

Since the construction and initialization of the product occur in two
separate transactions, it is possible that a third party might try to hijack
the process by inserting their own transaction to initialize the product.

> For instance, an adversary could invoke a different factory instance than indended by the user,

Reviewer: ```suggestion
For instance, an adversary could invoke a different factory instance than intended by the user,
```

as illustrated in the following sequence diagram:

.. image:: images/factory-adversary.svg
    :alt: sequence diagram showing how a third party might hijack a product

To prevent this possibility, the product checks in its ``initialize`` method that the invoker of the
transaction (i.e. the account that originated the transaction as a whole) is the same account as
created the product contract instance (i.e. the "owner"):

.. code-block:: Rust

        ensure_eq!(ctx.invoker(), ctx.owner(), ProductError::NotAuthorized);

With this check, the attack described above would result in failure for Adversary (because
the invoker Adversary does not match the owner User), but success for User:

.. image:: images/factory-adversary2.svg
    :alt: sequence diagram showing how a hijacking attempt fails

.. Note::
    Typically, it is wrong to use the invoker of a transaction for
    authorization, rather than the immediate caller. For instance, a user might
    invoke some untrusted smart contract, and expect it is not authorized to

>     transfer tokens she holds on another contract. If the token-holding contract

Reviewer: ```suggestion
    transfer tokens it holds on another contract. If the token-holding contract
```

    user intended, but the user would still be able to create another product
    and have the factory produce that correctly.

At this point, it just remains to initialize the state of the product:

.. code-block:: Rust

        // The index is supplied as a parameter by the factory.
        let index: u64 = ctx.parameter_cursor().get()?;
        // This endpoint should only be called by another smart contract, namely the
        // factory, which we record in the state.
        let factory = match ctx.sender() {
            Address::Contract(ca) => ca,
            _ => Err(ProductError::SenderIsAccountAddress)?,
        };
        // Initialize the state.
        let product = Product {
            index,
            factory,
        };
        *state = ProductState::Initialized(product);

.. Note::

    If a user invokes ``initialize`` directly (rather than through a factory contract) it will fail.
    This is because ``initialize`` checks that the immediate caller is a smart contract, and
    records the contract address in the state. The fact that this prevents a user from directly
    invoking ``initialize`` is incidental. The design intent is not to prevent the user from

>     badly initializing products (which they could also do by invoking a "BadFactory" as previously

Reviewer 3: ```suggestion
    initializing products badly (which they could also do by invoking a "BadFactory" as previously
```
